### PR TITLE
Improve CSV dialect detection

### LINKED ISSUE #100: CSV is not being parsed correctly and GT crashes analysing it
Trying to validate this file (for further import on OS) but don't see what's wrong in it and GT doesn't help.

[PEF2016_AC01.csv.zip](https://github.com/frictionlessdata/goodtables/files/268161/PEF2016_AC01.csv.zip)

### PR DESCRIPTION
**Pull request by georgiana-b**
It fixes #100, #98.
Since the CSV dialect detection is statistical, having a fixed value isn't suitable for all cases and the detection fails for very short or very long files.
This implementation starts sampling from 2 rows and continues to add until it finds a delimiter that is not (entirely) alaphanumeric or space. It is also the lightest I could come up with. If you have a better idea, please share.

### CODE REVIEW

**Comment by pwalsh** [comment]
@akariv WDYT? I say we merge and give it a try.

File: goodtables/compat.py [hunk 1]
```diff
@@ -30,22 +31,22 @@
...
-            dialect.delimiter = dialect.delimiter.encode('utf-8')
-            dialect.quotechar = dialect.quotechar.encode('utf-8')
-        except csv.Error:
-            dialect = csv.excel
-
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+        
+        while not re.findall('[^\w ]', dialect.delimiter): 
+            first_lines.extend(list(data.readline()))
```
Comment (akariv) [inline goodtables/compat.py]: why not use `append(` instead of `extend(list(`?

File: goodtables/compat.py [hunk 2]
```diff
@@ -30,22 +31,22 @@
...
+            dialect = detect_dialect(first_lines)
+            
         def iterenc_utf8(data):
             for line in data:
                 yield line.encode('utf-8')
 
-        iter = chain(first_lines, data)
+        data.seek(0)
```
Comment (akariv) [inline goodtables/compat.py]: why do you assume that data is seekable?
why was this change needed?
if it was, why keep first_lines?

Comment (pwalsh) [inline goodtables/compat.py]: the DataTable class ensures that what is passed here is a seekable stream. a bit unclear, I know.

File: goodtables/compat.py [hunk 3]
```diff
@@ -30,22 +31,22 @@
...
-            dialect = csv.Sniffer().sniff(''.join(first_lines))
-            dialect.delimiter = dialect.delimiter.encode('utf-8')
-            dialect.quotechar = dialect.quotechar.encode('utf-8')
-        except csv.Error:
-            dialect = csv.excel
-
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+        
+        while not re.findall('[^\w ]', dialect.delimiter):
```
Comment (akariv) [inline goodtables/compat.py]: better to use [`isspace`](https://docs.python.org/2/library/stdtypes.html#str.isspace) for this, no?

File: goodtables/compat.py [hunk 4]
```diff
@@ -67,23 +81,35 @@ def iterenc_utf8(data):
...
+    def csv_reader(data, header_index, **kwargs):
 
         def line_iterator(data):
             for line in data:
                 yield line
+
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+
+        while not re.findall('[^\w ]', dialect.delimiter):
```
Comment (akariv) [inline goodtables/compat.py]: see above

File: goodtables/compat.py [hunk 5]
```diff
@@ -67,23 +81,35 @@ def iterenc_utf8(data):
...
 
         def line_iterator(data):
             for line in data:
                 yield line
+
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+
+        while not re.findall('[^\w ]', dialect.delimiter): 
+            first_lines.extend(list(data.readline()))
```
Comment (akariv) [inline goodtables/compat.py]: see above

File: goodtables/compat.py [hunk 6]
```diff
@@ -67,23 +81,35 @@ def iterenc_utf8(data):
...
+            dialect = detect_dialect(first_lines)
+
+        data.seek(0)
```
Comment (akariv) [inline goodtables/compat.py]: see above

File: goodtables/compat.py (hunk 2 above)
Comment (akariv) [inline goodtables/compat.py]: Yes, but who guarantees that we should be processing the CSV from position 0?
When I made this change (seek -> chain) it wasn't the case. It was because we were using two csv readers - one to read the headers and then passing the stream to another csv reader to read the data. In that case, seeking to 0 would re-read the headers.

Therefore, I wouldn't seek to 0 - we can imagine a 'comment reader' that reads in the CSV until there are no more 'comment lines' (I remember seeing an issue about that once), or another who reads the 'BOM' character in the beginning of the stream. Then these processors pass the CSV reader the stream in the correct position to start, which is not 0.

Comment (pwalsh) [inline goodtables/compat.py]: yes, correct. that is what `header_index` is for. there are other places in code that seek using it. @georgiana-b

Comment (akariv) [inline goodtables/compat.py]: Is there any reason not to keep `iter = chain(first_lines, data)` ?

File: goodtables/compat.py (hunk 1 above)
Comment (georgiana-b) [inline goodtables/compat.py]: because appending a list to a list creates a 2 dimensional list: 
```
In [7]: test_list = [2,3,4]
In [8]: test_list.append([6])
In [9]: test_list
Out[9]: [2, 3, 4, [6]]

In [10]: test_list.extend([8])
In [11]: test_list
Out[11]: [2, 3, 4, [6], 8]
```

Comment (akariv) [inline goodtables/compat.py]: No, only if you append a list. But `data.readline()` is not a list.
```
In [5]: test_list = [2,3,4]
In [6]: test_list.append(6)
In [7]: test_list
Out[7]: [2, 3, 4, 6]
```

File: goodtables/compat.py (hunk 2 above)
Comment (georgiana-b) [inline goodtables/compat.py]: Adam, functions in DataTable (and maybe other places) expect to start from 0: https://github.com/frictionlessdata/goodtables/blob/master/goodtables/datatable/datatable.py#L209. I started sampling from the `header_index` because it influences the accuracy of the detection but if I then use `chain`, it will only include the lines from `header_index` downwards which conflicts with what other functions expect and leads to incorrect results.

File: goodtables/compat.py (hunk 3 above)
Comment (georgiana-b) [inline goodtables/compat.py]: `[^\w ]` matches anything that is not alphanumeric or whitespace.  `not str.isspace()` returns true for alphanumeric which is not ok (ex: `m` is not a valid delimiter).

File: goodtables/compat.py (hunk 1 above)
Comment (georgiana-b) [inline goodtables/compat.py]: You're right, there's no reason to convert that to a list. I'll change it.

File: goodtables/compat.py (hunk 3 above)
Comment (akariv) [inline goodtables/compat.py]: You're right. 
Still, running a regexp match is quite slow.
Matching like `delimiter.isalnum() || delimiter in [' ', '_']` is quicker and IMO more readable.

File: goodtables/compat.py (hunk 2 above)
Comment (akariv) [inline goodtables/compat.py]: If this is an assumption then it's a very recent assumption.

See this code from less than 2 months ago, https://github.com/frictionlessdata/goodtables/blob/519f49cc0373eaf30dad25e744a7a3ad29a4a637/goodtables/datatable/datatable.py#L63
`csv_reader` is assumed here to *not* change the stream position.

File: goodtables/compat.py (hunk 3 above)
Comment (georgiana-b) [inline goodtables/compat.py]: That leaves uncovered the case for multiple spaces (ex: a tab that is actually 4 spaces). I think it's better to have all cases covered than to be very fast in this case because that check won't execute many times.

Comment (akariv) [inline goodtables/compat.py]: I don't think the delimiter can be more than 1 character, but alright.
We can still remove the double negation, right?

Comment (georgiana-b) [inline goodtables/compat.py]: Right, I removed it.

File: goodtables/compat.py (hunk 2 above)
Comment (georgiana-b) [inline goodtables/compat.py]: It's not an assumption. This is how that same function looks these days: https://github.com/frictionlessdata/goodtables/blob/master/goodtables/datatable/datatable.py#L63. I don't understand what's bothering about this part.

Comment (akariv) [inline goodtables/compat.py]: You're changing the behaviour of this function - a behaviour that was once _assumed_ to be the exact opposite.
I want to make sure you checked that no one else in the code base is also assuming that the reader will start from the current position of the stream. 

Coming to think of it, who else in the code base needs the lines before the headers? Why not return an iterator that begins with the header row?

**Comment by pwalsh** [comment]
closing because GoodTables datatable is about to be replaced by [Tabulator](https://github.com/frictionlessdata/tabulator-py), rendering this out of date.
